File: apkprobe/rules.py
"""MASTG-aligned static checks over a parsed manifest.

Each check returns zero or more findings using the shared schema from
``scopeward.findings`` when available, falling back to a local-compatible
Finding so apkprobe also runs standalone. Checks map to MASVS controls /
MASTG test ids so results slot into a coverage matrix.
"""

from __future__ import annotations

from .manifest import AppManifest

try:  # prefer the suite-wide schema
    from scopeward.findings import Finding, Severity
    _HAVE_SCOPEWARD = True
except Exception:  # pragma: no cover - standalone fallback
    import enum
    from dataclasses import dataclass, field

    class Severity(enum.IntEnum):
        INFO = 0
        LOW = 1
        MEDIUM = 2
        HIGH = 3
        CRITICAL = 4

        @classmethod
        def parse(cls, v):
            return v if isinstance(v, cls) else cls[str(v).upper()]

    @dataclass
    class Finding:  # minimal compatible shape
        title: str
        severity: "Severity"
        target: str
        description: str = ""
        module: str = "apkprobe"
        masvs: str = ""
        mastg_test: str = ""
        evidence: str = ""
        metadata: dict = field(default_factory=dict)

        def to_dict(self):
            d = dict(self.__dict__)
            d["severity"] = Severity.parse(self.severity).name
            return d

    _HAVE_SCOPEWARD = False


# Permissions worth surfacing (subset of Android's "dangerous"/sensitive set)
SENSITIVE_PERMISSIONS = {
    "android.permission.READ_SMS",
    "android.permission.SEND_SMS",
    "android.permission.READ_CONTACTS",
    "android.permission.ACCESS_FINE_LOCATION",
    "android.permission.RECORD_AUDIO",
    "android.permission.READ_EXTERNAL_STORAGE",
    "android.permission.WRITE_EXTERNAL_STORAGE",
    "android.permission.READ_PHONE_STATE",
    "android.permission.CAMERA",
    "android.permission.REQUEST_INSTALL_PACKAGES",
}


def analyze_manifest(m: AppManifest) -> list:
    target = f"android:{m.package}" if m.package else "android:unknown"
    findings: list = []

    def add(title, sev, masvs, mastg, evidence, desc=""):
        findings.append(Finding(
            title=title, severity=Severity.parse(sev), target=target,
            module="apkprobe", masvs=masvs, mastg_test=mastg,
            evidence=evidence, description=desc,
        ))

    if m.debuggable:
        add("Application is debuggable", "HIGH", "MASVS-RESILIENCE-2", "MASTG-TEST-0026",
            "android:debuggable=\"true\"",
            "A debuggable build lets anyone attach a debugger and inspect/modify the running app.")

    if m.allow_backup:
        add("ADB backup allowed", "MEDIUM", "MASVS-STORAGE-2", "MASTG-TEST-0009",
            "android:allowBackup=\"true\"",
            "App data can be extracted via 'adb backup' on a debuggable-or-rooted device.")

    if m.uses_cleartext_traffic is True:
        add("Cleartext (HTTP) traffic permitted", "HIGH", "MASVS-NETWORK-1", "MASTG-TEST-0019",
            "android:usesCleartextTraffic=\"true\"",
            "App allows unencrypted network traffic, exposing data to interception.")

    if m.target_sdk and m.target_sdk >= 24 and not m.network_security_config \
            and m.uses_cleartext_traffic is not False:
        add("No Network Security Config", "LOW", "MASVS-NETWORK-2", "MASTG-TEST-0020",
            f"targetSdk={m.target_sdk}, no android:networkSecurityConfig",
            "No NSC means default trust + no pinning policy; consider an explicit config.")

    for c in m.components:
        if c.exported and not c.has_permission:
            sev = "MEDIUM" if c.intent_filters else "LOW"
            add(f"Exported {c.kind} without permission: {c.name}", sev,
                "MASVS-PLATFORM-1", "MASTG-TEST-0024",
                f"{c.kind} {c.name} exported=true, permission=none, intent-filters={c.intent_filters}",
                "Exported component reachable by other apps with no permission guard.")

    for perm in sorted(set(m.permissions) & SENSITIVE_PERMISSIONS):
        add(f"Sensitive permission requested: {perm}", "INFO",
            "MASVS-PLATFORM-1", "MASTG-TEST-0024", perm,
            "Verify this permission is actually required and justified.")

    if m.min_sdk and m.min_sdk < 24:
        add(f"Low minSdkVersion ({m.min_sdk})", "LOW", "MASVS-RESILIENCE-1", "",
            f"minSdkVersion={m.min_sdk}",
            "Supporting old Android versions inherits their unpatched platform vulnerabilities.")

    return findings

File: ports/go/rules.go
// Command apkprobe-rules is a Go port of apkprobe's core MASVS/MASTG manifest
// rule engine (see apkprobe/rules.py). It consumes a normalized manifest JSON
// and emits the same findings as the Python reference. Defensive, offline, no
// network or device access.
package main

import "sort"

// Component mirrors apkprobe.manifest.Component.
type Component struct {
	Kind          string `json:"kind"`
	Name          string `json:"name"`
	Exported      bool   `json:"exported"`
	HasPermission bool   `json:"has_permission"`
	IntentFilters int    `json:"intent_filters"`
}

// Manifest mirrors apkprobe.manifest.AppManifest (normalized).
// UsesCleartextTraffic is a *bool: nil = unset.
type Manifest struct {
	Package               string     `json:"package"`
	MinSDK                int        `json:"min_sdk"`
	TargetSDK             int        `json:"target_sdk"`
	Debuggable            bool       `json:"debuggable"`
	AllowBackup           bool       `json:"allow_backup"`
	UsesCleartextTraffic  *bool      `json:"uses_cleartext_traffic"`
	NetworkSecurityConfig string     `json:"network_security_config"`
	Permissions           []string   `json:"permissions"`
	Components            []Component `json:"components"`
}

// Finding mirrors the subset of fields the ports agree on.
type Finding struct {
	Title     string `json:"title"`
	Severity  string `json:"severity"`
	MASVS     string `json:"masvs"`
	MASTGTest string `json:"mastg_test"`
	Evidence  string `json:"evidence"`
}

// SensitivePermissions matches rules.SENSITIVE_PERMISSIONS.
var SensitivePermissions = map[string]bool{
	"android.permission.READ_SMS":                true,
	"android.permission.SEND_SMS":                true,
	"android.permission.READ_CONTACTS":           true,
	"android.permission.ACCESS_FINE_LOCATION":    true,
	"android.permission.RECORD_AUDIO":            true,
	"android.permission.READ_EXTERNAL_STORAGE":   true,
	"android.permission.WRITE_EXTERNAL_STORAGE":  true,
	"android.permission.READ_PHONE_STATE":        true,
	"android.permission.CAMERA":                  true,
	"android.permission.REQUEST_INSTALL_PACKAGES": true,
}

// AnalyzeManifest runs the MASVS/MASTG checks over m, in the same order as the
// Python reference.
func AnalyzeManifest(m Manifest) []Finding {
	var out []Finding
	add := func(title, sev, masvs, mastg, ev string) {
		out = append(out, Finding{Title: title, Severity: sev, MASVS: masvs, MASTGTest: mastg, Evidence: ev})
	}

	if m.Debuggable {
		add("Application is debuggable", "HIGH", "MASVS-RESILIENCE-2", "MASTG-TEST-0026",
			`android:debuggable="true"`)
	}
	if m.AllowBackup {
		add("ADB backup allowed", "MEDIUM", "MASVS-STORAGE-2", "MASTG-TEST-0009",
			`android:allowBackup="true"`)
	}
	if m.UsesCleartextTraffic != nil && *m.UsesCleartextTraffic {
		add("Cleartext (HTTP) traffic permitted", "HIGH", "MASVS-NETWORK-1", "MASTG-TEST-0019",
			`android:usesCleartextTraffic="true"`)
	}
	if m.TargetSDK >= 24 && m.NetworkSecurityConfig == "" &&
		!(m.UsesCleartextTraffic != nil && !*m.UsesCleartextTraffic) {
		add("No Network Security Config", "LOW", "MASVS-NETWORK-2", "MASTG-TEST-0020",
			"targetSdk="+itoa(m.TargetSDK)+", no android:networkSecurityConfig")
	}
	for _, c := range m.Components {
		if c.Exported && !c.HasPermission {
			sev := "LOW"
			if c.IntentFilters > 0 {
				sev = "MEDIUM"
			}
			add("Exported "+c.Kind+" without permission: "+c.Name, sev,
				"MASVS-PLATFORM-1", "MASTG-TEST-0024",
				c.Kind+" "+c.Name+" exported=true, permission=none, intent-filters="+itoa(c.IntentFilters))
		}
	}
	// sorted(set(permissions) & SENSITIVE)
	seen := map[string]bool{}
	var sensitive []string
	for _, p := range m.Permissions {
		if SensitivePermissions[p] && !seen[p] {
			seen[p] = true
			sensitive = append(sensitive, p)
		}
	}
	sort.Strings(sensitive)
	for _, p := range sensitive {
		add("Sensitive permission requested: "+p, "INFO",
			"MASVS-PLATFORM-1", "MASTG-TEST-0024", p)
	}
	if m.MinSDK > 0 && m.MinSDK < 24 {
		add("Low minSdkVersion ("+itoa(m.MinSDK)+")", "LOW", "MASVS-RESILIENCE-1", "",
			"minSdkVersion="+itoa(m.MinSDK))
	}
	return out
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [20]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}
